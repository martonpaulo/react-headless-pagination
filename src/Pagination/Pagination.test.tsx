import React from "react";
import { FiArrowLeft, FiArrowRight } from "react-icons/fi";
import { render, screen, fireEvent } from "@testing-library/react";
import { Pagination } from "./Pagination";
import { IPaginationProps } from "../types/types";

const setupPagination = ({
  currentPage = 5,
  setCurrentPage = () => {},
  className = "",
  truncableText = "...",
  truncableClassName = "",
  totalPages = 10,
  edgePageCount = 1,
  middlePagesSiblingCount = 2,
  children = (
    <>
      <Pagination.PrevButton dataTestId="prev">Previous</Pagination.PrevButton>

      <nav className="flex justify-center flex-grow">
        <ul className="flex items-center">
          <Pagination.PageButton
            dataTestIdActive="active-page-button"
            dataTestIdInactive="inactive-page-button"
            activeClassName=""
            inactiveClassName=""
            className=""
          />
        </ul>
      </nav>

      <Pagination.NextButton dataTestId="next">Next</Pagination.NextButton>
    </>
  ),
}: Partial<IPaginationProps>) =>
  render(
    <Pagination
      currentPage={currentPage}
      setCurrentPage={setCurrentPage}
      className={className}
      truncableText={truncableText}
      truncableClassName={truncableClassName}
      totalPages={totalPages}
      edgePageCount={edgePageCount}
      middlePagesSiblingCount={middlePagesSiblingCount}
    >
      {children}
    </Pagination>,
  );

describe("Pagination", () => {
  it("renders correctly with basic setup of prevButton, pageButton and nextButton", () => {
    const { asFragment } = setupPagination({});

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with different truncable text", () => {
    const { asFragment } = setupPagination({
      truncableText: "---",
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with advanced setup of prevButton, pageButton and nextButton using Tailwind", () => {
    const { asFragment } = setupPagination({
      className: "flex items-center w-full h-10 text-sm select-none",
      truncableClassName: "w-10 px-0.5 text-center",
      children: (
        <>
          <Pagination.PrevButton className="flex items-center mr-2 text-gray-500 cursor-pointer hover:text-gray-600 dark:hover:text-gray-200 focus:outline-none">
            <FiArrowLeft size={20} className="mr-3" />
            Previous
          </Pagination.PrevButton>

          <nav className="flex justify-center flex-grow">
            <ul className="flex items-center">
              <Pagination.PageButton
                activeClassName="bg-primary-50 dark:bg-opacity-0 text-primary-600 dark:text-white"
                inactiveClassName="text-gray-500"
                className={
                  "flex items-center justify-center h-10 w-10 rounded-full cursor-pointer"
                }
              />
            </ul>
          </nav>

          <Pagination.NextButton className="flex items-center mr-2 text-gray-500 opacity-50 hover:text-gray-600 dark:hover:text-gray-200 focus:outline-none">
            Next
            <FiArrowRight size={20} className="ml-3" />
          </Pagination.NextButton>
        </>
      ),
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with custom page button", () => {
    const { asFragment } = setupPagination({
      children: (
        <>
          <Pagination.PrevButton>Previous</Pagination.PrevButton>

          <div className="flex items-center justify-center flex-grow">
            <Pagination.PageButton
              as={<span />}
              dataTestIdActive="page-button-custom"
            />
          </div>

          <Pagination.NextButton>Next</Pagination.NextButton>
        </>
      ),
    });

    expect(asFragment()).toMatchSnapshot();

    const headingElement = screen.getByTestId("page-button-custom");
    expect(headingElement.tagName.toLowerCase()).toEqual("span");
  });

  it("renders correctly with extra page button props", () => {
    const { asFragment } = setupPagination({
      children: (
        <>
          <Pagination.PrevButton>Previous</Pagination.PrevButton>

          <div className="flex items-center justify-center flex-grow">
            <Pagination.PageButton
              as={<span />}
              dataTestIdActive="page-button-extra-props"
              renderExtraProps={(page) => ({
                "aria-label": `Go to page ${page}`,
              })}
            />
          </div>

          <Pagination.NextButton>Next</Pagination.NextButton>
        </>
      ),
    });

    expect(asFragment()).toMatchSnapshot();

    const pageBtnElem = screen.getByTestId("page-button-extra-props");
    expect(pageBtnElem.getAttribute("aria-label")).toContain("Go to page");
  });

  it("clicking on previous calls mockSetCurrentPage with currentPage - 1", () => {
    const mockSetCurrentPage = jest.fn();

    setupPagination({
      currentPage: 4,
      totalPages: 5,
      setCurrentPage: mockSetCurrentPage,
    });

    expect(screen.getByTestId("active-page-button").textContent).toEqual("5");

    screen.getByTestId("prev").click();

    expect(mockSetCurrentPage).toHaveBeenCalledWith(3);
  });

  it("clicking on next calls mockSetCurrentPage with currentPage + 1", () => {
    const mockSetCurrentPage = jest.fn();

    setupPagination({
      currentPage: 1,
      totalPages: 5,
      setCurrentPage: mockSetCurrentPage,
    });

    expect(screen.getByTestId("active-page-button").textContent).toEqual("2");

    screen.getByTestId("next").click();

    expect(mockSetCurrentPage).toHaveBeenCalledWith(2);
  });

  it("clicking on current page calls mockSetCurrentPage with currentPage", () => {
    const mockSetCurrentPage = jest.fn();

    setupPagination({
      currentPage: 5,
      totalPages: 10,
      setCurrentPage: mockSetCurrentPage,
    });

    expect(screen.getByTestId("active-page-button").textContent).toEqual("6");

    screen.getByTestId("inactive-page-button-10").click();

    expect(mockSetCurrentPage).toHaveBeenCalledWith(9);
  });

  it("pressing enter when focused on page button fires setCurrentPage", () => {
    const mockSetCurrentPage = jest.fn();

    setupPagination({
      currentPage: 5,
      totalPages: 10,
      setCurrentPage: mockSetCurrentPage,
    });

    const pageButton4 = screen.getByText("4");
    pageButton4.focus();

    fireEvent.keyPress(pageButton4, {
      key: "Enter",
      code: "Enter",
      charCode: 13,
    });

    expect(mockSetCurrentPage).toHaveBeenCalledWith(3);
  });
});
