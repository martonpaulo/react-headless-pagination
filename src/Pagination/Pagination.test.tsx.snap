// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly with advanced setup of prevButton, pageButton and nextButton using Tailwind 1`] = `
<DocumentFragment>
  <div
    class="flex items-center w-full h-10 text-sm select-none"
  >
    <button
      class="flex items-center mr-2 text-gray-500 cursor-pointer hover:text-gray-600 dark:hover:text-gray-200 focus:outline-none"
      tabindex="0"
    >
      <svg
        class="mr-3"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="19"
          x2="5"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 19 5 12 12 5"
        />
      </svg>
      Previous
    </button>
    <nav
      class="flex justify-center flex-grow"
    >
      <ul
        class="flex items-center"
      >
        <li>
          <a
            class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
            tabindex="0"
          >
            1
          </a>
        </li>
        <li
          class="w-10 px-0.5 text-center"
        >
          ...
        </li>
        <li>
          <a
            class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
            tabindex="0"
          >
            4
          </a>
        </li>
        <li>
          <a
            class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
            tabindex="0"
          >
            5
          </a>
        </li>
        <li>
          <a
            class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer bg-primary-50 dark:bg-opacity-0 text-primary-600 dark:text-white"
            tabindex="0"
          >
            6
          </a>
        </li>
        <li>
          <a
            class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
            tabindex="0"
          >
            7
          </a>
        </li>
        <li>
          <a
            class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
            tabindex="0"
          >
            8
          </a>
        </li>
        <li
          class="w-10 px-0.5 text-center"
        >
          ...
        </li>
        <li>
          <a
            class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
            tabindex="0"
          >
            10
          </a>
        </li>
      </ul>
    </nav>
    <button
      class="flex items-center mr-2 text-gray-500 opacity-50 hover:text-gray-600 dark:hover:text-gray-200 focus:outline-none"
      tabindex="0"
    >
      Next
      <svg
        class="ml-3"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="5"
          x2="19"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 5 19 12 12 19"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Pagination renders correctly with basic setup of prevButton, pageButton and nextButton 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <button
      class=""
      data-testid="prev"
      tabindex="0"
    >
      Previous
    </button>
    <nav
      class="flex justify-center flex-grow"
    >
      <ul
        class="flex items-center"
      >
        <li>
          <a
            class=""
            data-testid="inactive-page-button-1"
            tabindex="0"
          >
            1
          </a>
        </li>
        <li>
          ...
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-4"
            tabindex="0"
          >
            4
          </a>
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-5"
            tabindex="0"
          >
            5
          </a>
        </li>
        <li>
          <a
            class=""
            data-testid="active-page-button"
            tabindex="0"
          >
            6
          </a>
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-7"
            tabindex="0"
          >
            7
          </a>
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-8"
            tabindex="0"
          >
            8
          </a>
        </li>
        <li>
          ...
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-10"
            tabindex="0"
          >
            10
          </a>
        </li>
      </ul>
    </nav>
    <button
      class=""
      data-testid="next"
      tabindex="0"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;

exports[`Pagination renders correctly with custom page button 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <button
      class=""
      tabindex="0"
    >
      Previous
    </button>
    <div
      class="flex items-center justify-center flex-grow"
    >
      <li>
        <span
          class=""
          data-testid="undefined-1"
          tabindex="0"
        >
          1
        </span>
      </li>
      <li>
        ...
      </li>
      <li>
        <span
          class=""
          data-testid="undefined-4"
          tabindex="0"
        >
          4
        </span>
      </li>
      <li>
        <span
          class=""
          data-testid="undefined-5"
          tabindex="0"
        >
          5
        </span>
      </li>
      <li>
        <span
          class=""
          data-testid="page-button-custom"
          tabindex="0"
        >
          6
        </span>
      </li>
      <li>
        <span
          class=""
          data-testid="undefined-7"
          tabindex="0"
        >
          7
        </span>
      </li>
      <li>
        <span
          class=""
          data-testid="undefined-8"
          tabindex="0"
        >
          8
        </span>
      </li>
      <li>
        ...
      </li>
      <li>
        <span
          class=""
          data-testid="undefined-10"
          tabindex="0"
        >
          10
        </span>
      </li>
    </div>
    <button
      class=""
      tabindex="0"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;

exports[`Pagination renders correctly with different truncable text 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <button
      class=""
      data-testid="prev"
      tabindex="0"
    >
      Previous
    </button>
    <nav
      class="flex justify-center flex-grow"
    >
      <ul
        class="flex items-center"
      >
        <li>
          <a
            class=""
            data-testid="inactive-page-button-1"
            tabindex="0"
          >
            1
          </a>
        </li>
        <li>
          ---
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-4"
            tabindex="0"
          >
            4
          </a>
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-5"
            tabindex="0"
          >
            5
          </a>
        </li>
        <li>
          <a
            class=""
            data-testid="active-page-button"
            tabindex="0"
          >
            6
          </a>
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-7"
            tabindex="0"
          >
            7
          </a>
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-8"
            tabindex="0"
          >
            8
          </a>
        </li>
        <li>
          ---
        </li>
        <li>
          <a
            class=""
            data-testid="inactive-page-button-10"
            tabindex="0"
          >
            10
          </a>
        </li>
      </ul>
    </nav>
    <button
      class=""
      data-testid="next"
      tabindex="0"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;

exports[`Pagination renders correctly with extra page button props 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <button
      class=""
      tabindex="0"
    >
      Previous
    </button>
    <div
      class="flex items-center justify-center flex-grow"
    >
      <li>
        <span
          aria-label="Go to page 1"
          class=""
          data-testid="undefined-1"
          tabindex="0"
        >
          1
        </span>
      </li>
      <li>
        ...
      </li>
      <li>
        <span
          aria-label="Go to page 4"
          class=""
          data-testid="undefined-4"
          tabindex="0"
        >
          4
        </span>
      </li>
      <li>
        <span
          aria-label="Go to page 5"
          class=""
          data-testid="undefined-5"
          tabindex="0"
        >
          5
        </span>
      </li>
      <li>
        <span
          aria-label="Go to page 6"
          class=""
          data-testid="page-button-extra-props"
          tabindex="0"
        >
          6
        </span>
      </li>
      <li>
        <span
          aria-label="Go to page 7"
          class=""
          data-testid="undefined-7"
          tabindex="0"
        >
          7
        </span>
      </li>
      <li>
        <span
          aria-label="Go to page 8"
          class=""
          data-testid="undefined-8"
          tabindex="0"
        >
          8
        </span>
      </li>
      <li>
        ...
      </li>
      <li>
        <span
          aria-label="Go to page 10"
          class=""
          data-testid="undefined-10"
          tabindex="0"
        >
          10
        </span>
      </li>
    </div>
    <button
      class=""
      tabindex="0"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;
